src/Microsoft.AspNetCore.Routing/EndpointConstraints/EndpointSelector.cs

```diff
@@ -0,0 +1,442 @@
+                Volatile.Write(ref _cache, cache);
+                return cache;
+            }
+        }
+
+        public IReadOnlyList<Endpoint> SelectCandidates(RouteContext context)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            var cache = Current;
+
+            // The Cache works based on a string[] of the route values in a pre-calculated order. This code extracts
+            // those values in the correct order.
+            var keys = cache.RouteKeys;
+            var values = new string[keys.Length];
+            for (var i = 0; i < keys.Length; i++)
+            {
+                context.RouteData.Values.TryGetValue(keys[i], out object value);
+
+                if (value != null)
+                {
+                    values[i] = value as string ?? Convert.ToString(value);
+                }
+            }
+
+            if (cache.OrdinalEntries.TryGetValue(values, out var matchingRouteValues) ||
+                cache.OrdinalIgnoreCaseEntries.TryGetValue(values, out matchingRouteValues))
+            {
+                Debug.Assert(matchingRouteValues != null);
+                return matchingRouteValues;
+            }
+
+            //_logger.NoEndpointsMatched(context.RouteData.Values);
+            return EmptyEndpoints;
+        }
+
+        public Endpoint SelectBestCandidate(HttpContext context, IReadOnlyList<Endpoint> candidates)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            if (candidates == null)
+            {
+                throw new ArgumentNullException(nameof(candidates));
+            }
+
+            var finalMatches = EvaluateEndpointConstraints(context, candidates);
+
+            if (finalMatches == null || finalMatches.Count == 0)
+            {
+                return null;
+            }
+            else if (finalMatches.Count == 1)
+            {
+                var selectedEndpoint = finalMatches[0];
+
+                return selectedEndpoint;
+            }
+            else
+            {
+                var endpointNames = string.Join(
+                    Environment.NewLine,
+                    finalMatches.Select(a => a.DisplayName));
+
+                Log.MatchAmbiguous(_logger, context, finalMatches);
+
+                var message = Resources.FormatAmbiguousEndpoints(
+                    Environment.NewLine,
+                    string.Join(Environment.NewLine, endpointNames));
+
+                throw new AmbiguousMatchException(message);
+            }
+        }
+
+        private IReadOnlyList<Endpoint> EvaluateEndpointConstraints(
+            HttpContext context,
+            IReadOnlyList<Endpoint> endpoints)
+        {
+            var candidates = new List<EndpointSelectorCandidate>();
+
+            // Perf: Avoid allocations
```

Comment:
I'm really satisified that we're finally doing perf work on this code. It's adorable that once it was *worth a comment* to explain that I was using a for-loop on purpose. These were very early days for the product.

Reply:
I suspect around ASP.NET 5 ***beta 8*** you were probably all losing your minds 😋